### ISSUE TITLE
Accessibility traversal within a scrollable not working correctly

### ISSUE THREAD

**Issue by lauweijie** [flutter/flutter#55561]
Internal: b/154912334

Accessibility traversal is not working correctly for scrollable when the child's height is larger than the scrollable viewport.

## Steps to Reproduce

There are no issues navigating through boxes 1 to 6 by swiping left and right on TalkBack. To reach "Box 9", the user has to use the two-finger scroll gesture and cannot rely on the natural accessibility traversal (see [video](https://github.com/flutter/flutter/files/4529521/scrollable-a11y-traversal.mp4.zip)).

```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: Text('Test app')),
        body: SingleChildScrollView(
          child: Column(
            children: [
              _buildContainer(text: 'Box 1', height: 250, color: Colors.red),
              _buildContainer(text: 'Box 2', height: 250, color: Colors.cyan),
              _buildContainer(text: 'Box 3', height: 250, color: Colors.red),
              _buildContainer(text: 'Box 4', height: 250, color: Colors.cyan),
              _buildContainer(text: 'Box 5', height: 250, color: Colors.red),
              _buildContainer(text: 'Box 6', height: 250, color: Colors.cyan),
              _buildContainer(text: 'Box 7', height: 1000, color: Colors.amber),
              _buildContainer(text: 'Box 8', height: 1000, color: Colors.brown),
              _buildContainer(text: 'Box 9', height: 1000, color: Colors.grey),
            ],
          ),
        ),
      ),
    );
  }
}

Widget _buildContainer({double height, Color color, String text}) => Container(
  color: color,
  height: height,
  width: double.infinity,
  alignment: Alignment.center,
  child: Text(text, style: TextStyle(fontSize: 24)),
);
```

**Expected results:**
Swiping right in TalkBack should navigate through Boxes 1 - 9.

**Actual results:**
Swiping right after Box 7 doesn't move accessibility focus to Box 8.

**Comment by mehmetf** [comment]
@goderbauer I somehow recall seeing a similar bug to this. What is the expected behavior for accessibility traversal if list is lazily built. Should it scroll and build widgets? or should it only traverse whatever is visible?

**Comment by goderbauer** [comment]
In theory, you should be able to traverse Boxes 1 - 9.  However, I can reproduce that this is not working in this example. I suspect the culprit here are the unusually large boxes in combination with how TalkBack works: When the last visible element of a list has focus and the user swipes to go to the next element, TalkBack checks whether the list is scrollable. If it is, TalkBack tells Flutter to scroll the list and then moves the focus to the newly uncovered next element. However, if the space between focusable elements is too large (as in this example) it's possible that one scroll action is not enough to expose the next focusable element.

I don't have a great idea how to fix or work around this. One solution would obviously be to have smaller containers to ensure that you can expose the next focusable item with one scroll action. However, depending on the app, this may not be feasible, of course.

**Comment by mehmetf** [comment]
Is there a precise definition of `one scroll action`?

**Comment by goderbauer** [comment]
TalkBack dispatches "one scroll action" when you swipe left and then right in one movement (for down) and when you swipe right and then left in one movement (for up). This triggers Flutter's `OnSrollUp/Down` [1] callback. For a scrollable that one is currently wired up to scroll the scrollable by 80% of it's height:

https://github.com/flutter/flutter/blob/dd88b20435c1eb9b837bb4aa8b0a84aa0825530d/packages/flutter/lib/src/rendering/proxy_box.dart#L3521-L3529


[1] https://master-api.flutter.dev/flutter/semantics/SemanticsConfiguration/onScrollUp.html

**Comment by mehmetf** [comment]
Native Android apps also behave like Flutter so this is being closed as "working as intended".

The real solution, as @goderbauer says, is to make content smaller and thus a11y compatible.
